File: frontend/community_qa_app/lib/models/space.dart
import 'package:json_annotation/json_annotation.dart';

part 'space.g.dart';

@JsonSerializable()
class Space {
  final String id;
  final String name;
  final String slug;
  final String description;
  final String? rules;
  final String? icon;
  final String? banner;
  final String color;
  @JsonKey(defaultValue: false)
  final bool isPublic;
  @JsonKey(defaultValue: false)
  final bool isActive;
  @JsonKey(defaultValue: false)
  final bool allowQuestions;
  @JsonKey(defaultValue: false)
  final bool allowDiscussions;
  @JsonKey(defaultValue: false)
  final bool requireApproval;
  @JsonKey(defaultValue: false)
  final bool allowAnonymous;
  final int memberCount;
  final int postCount;
  final int questionCount;
  final int discussionCount;
  final List<String>? defaultTags;
  final String owner;
  final List<SpaceModerator>? moderators;
  final List<SpaceCategory>? categories;
  final SpaceStats? stats;
  final DateTime? lastActivity;
  final DateTime createdAt;
  final DateTime updatedAt;

  Space({
    required this.id,
    required this.name,
    required this.slug,
    required this.description,
    this.rules,
    this.icon,
    this.banner,
    required this.color,
    required this.isPublic,
    required this.isActive,
    required this.allowQuestions,
    required this.allowDiscussions,
    required this.requireApproval,
    required this.allowAnonymous,
    required this.memberCount,
    required this.postCount,
    required this.questionCount,
    required this.discussionCount,
    this.defaultTags,
    required this.owner,
    this.moderators,
    this.categories,
    this.stats,
    this.lastActivity,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Space.fromJson(Map<String, dynamic> json) => _$SpaceFromJson(json);
  Map<String, dynamic> toJson() => _$SpaceToJson(this);

  String get iconUrl {
    if (icon != null && icon!.isNotEmpty) {
      if (icon!.startsWith('http')) {
        return icon!;
      }
      return 'http://localhost:3000/uploads/spaces/icons/$icon';
    }
    return '';
  }

  String get bannerUrl {
    if (banner != null && banner!.isNotEmpty) {
      if (banner!.startsWith('http')) {
        return banner!;
      }
      return 'http://localhost:3000/uploads/spaces/banners/$banner';
    }
    return '';
  }

  Space copyWith({
    String? id,
    String? name,
    String? slug,
    String? description,
    String? rules,
    String? icon,
    String? banner,
    String? color,
    bool? isPublic,
    bool? isActive,
    bool? allowQuestions,
    bool? allowDiscussions,
    bool? requireApproval,
    bool? allowAnonymous,
    int? memberCount,
    int? postCount,
    int? questionCount,
    int? discussionCount,
    List<String>? defaultTags,
    String? owner,
    List<SpaceModerator>? moderators,
    List<SpaceCategory>? categories,
    SpaceStats? stats,
    DateTime? lastActivity,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Space(
      id: id ?? this.id,
      name: name ?? this.name,
      slug: slug ?? this.slug,
      description: description ?? this.description,
      rules: rules ?? this.rules,
      icon: icon ?? this.icon,
      banner: banner ?? this.banner,
      color: color ?? this.color,
      isPublic: isPublic ?? this.isPublic,
      isActive: isActive ?? this.isActive,
      allowQuestions: allowQuestions ?? this.allowQuestions,
      allowDiscussions: allowDiscussions ?? this.allowDiscussions,
      requireApproval: requireApproval ?? this.requireApproval,
      allowAnonymous: allowAnonymous ?? this.allowAnonymous,
      memberCount: memberCount ?? this.memberCount,
      postCount: postCount ?? this.postCount,
      questionCount: questionCount ?? this.questionCount,
      discussionCount: discussionCount ?? this.discussionCount,
      defaultTags: defaultTags ?? this.defaultTags,
      owner: owner ?? this.owner,
      moderators: moderators ?? this.moderators,
      categories: categories ?? this.categories,
      stats: stats ?? this.stats,
      lastActivity: lastActivity ?? this.lastActivity,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}

@JsonSerializable()
class SpaceModerator {
  final String user;
  final String addedBy;
  final DateTime addedAt;

  SpaceModerator({
    required this.user,
    required this.addedBy,
    required this.addedAt,
  });

  factory SpaceModerator.fromJson(Map<String, dynamic> json) => _$SpaceModeratorFromJson(json);
  Map<String, dynamic> toJson() => _$SpaceModeratorToJson(this);
}

@JsonSerializable()
class SpaceCategory {
  final String name;
  final String? description;
  final String color;
  final int order;

  SpaceCategory({
    required this.name,
    this.description,
    required this.color,
    required this.order,
  });

  factory SpaceCategory.fromJson(Map<String, dynamic> json) => _$SpaceCategoryFromJson(json);
  Map<String, dynamic> toJson() => _$SpaceCategoryToJson(this);
}

@JsonSerializable()
class SpaceStats {
  final int totalViews;
  final int totalVotes;
  final double averageResponseTime;
  final int unansweredQuestions;

  SpaceStats({
    required this.totalViews,
    required this.totalVotes,
    required this.averageResponseTime,
    required this.unansweredQuestions,
  });

  factory SpaceStats.fromJson(Map<String, dynamic> json) => _$SpaceStatsFromJson(json);
  Map<String, dynamic> toJson() => _$SpaceStatsToJson(this);
}

File: src/models/Space.js
const mongoose = require('mongoose');
const { SPACE } = require('../utils/constants');

const spaceSchema = new mongoose.Schema({
  // Basic information
  name: {
    type: String,
    required: [true, 'Space name is required'],
    trim: true,
    minlength: [2, 'Space name must be at least 2 characters long'],
    maxlength: [SPACE.MAX_NAME_LENGTH, `Space name cannot exceed ${SPACE.MAX_NAME_LENGTH} characters`],
    unique: true
  },
  slug: {
    type: String,
    required: true,
    unique: true,
    lowercase: true,
    trim: true
  },
  description: {
    type: String,
    required: [true, 'Space description is required'],
    trim: true,
    maxlength: [SPACE.MAX_DESCRIPTION_LENGTH, `Description cannot exceed ${SPACE.MAX_DESCRIPTION_LENGTH} characters`]
  },
  rules: {
    type: String,
    maxlength: [SPACE.MAX_RULES_LENGTH, `Rules cannot exceed ${SPACE.MAX_RULES_LENGTH} characters`],
    default: ''
  },

  // Visual elements
  icon: {
    type: String,
    default: ''
  },
  banner: {
    type: String,
    default: ''
  },
  color: {
    type: String,
    default: '#007bff',
    match: [/^#[0-9A-F]{6}$/i, 'Color must be a valid hex color']
  },

  // Settings
  isPublic: {
    type: Boolean,
    default: true
  },
  isActive: {
    type: Boolean,
    default: true
  },
  allowQuestions: {
    type: Boolean,
    default: true
  },
  allowDiscussions: {
    type: Boolean,
    default: true
  },
  requireApproval: {
    type: Boolean,
    default: false
  },
  allowAnonymous: {
    type: Boolean,
    default: false
  },

  // Membership
  memberCount: {
    type: Number,
    default: 0
  },
  postCount: {
    type: Number,
    default: 0
  },
  questionCount: {
    type: Number,
    default: 0
  },
  discussionCount: {
    type: Number,
    default: 0
  },

  // Default tags for this space
  defaultTags: [{
    type: String,
    trim: true,
    maxlength: [SPACE.MAX_TAGS_PER_SPACE, `Cannot have more than ${SPACE.MAX_TAGS_PER_SPACE} default tags`]
  }],

  // Moderators and owners
  owner: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  moderators: [{
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    addedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User'
    },
    addedAt: {
      type: Date,
      default: Date.now
    }
  }],

  // Categories and topics
  categories: [{
    name: {
      type: String,
      required: true,
      trim: true
    },
    description: {
      type: String,
      trim: true
    },
    color: {
      type: String,
      default: '#6c757d'
    },
    order: {
      type: Number,
      default: 0
    }
  }],

  // Statistics
  stats: {
    totalViews: {
      type: Number,
      default: 0
    },
    totalVotes: {
      type: Number,
      default: 0
    },
    averageResponseTime: {
      type: Number, // in hours
      default: 0
    },
    unansweredQuestions: {
      type: Number,
      default: 0
    }
  },

  // Activity tracking
  lastActivity: {
    type: Date,
    default: Date.now
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  updatedAt: {
    type: Date,
    default: Date.now
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Indexes
spaceSchema.index({ slug: 1 });
spaceSchema.index({ name: 'text', description: 'text' });
spaceSchema.index({ isPublic: 1, isActive: 1 });
spaceSchema.index({ memberCount: -1 });
spaceSchema.index({ postCount: -1 });
spaceSchema.index({ createdAt: -1 });
spaceSchema.index({ lastActivity: -1 });

// Virtual for space URL
spaceSchema.virtual('url').get(function() {
  return `/spaces/${this.slug}`;
});

// Virtual for icon URL
spaceSchema.virtual('iconUrl').get(function() {
  if (this.icon) {
    return this.icon.startsWith('http') ? this.icon : `/uploads/spaces/icons/${this.icon}`;
  }
  return '';
});

// Virtual for banner URL
spaceSchema.virtual('bannerUrl').get(function() {
  if (this.banner) {
    return this.banner.startsWith('http') ? this.banner : `/uploads/spaces/banners/${this.banner}`;
  }
  return '';
});

// Pre-save middleware to generate slug
spaceSchema.pre('save', function(next) {
  if (this.isModified('name')) {
    this.slug = this.name
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, '-')
      .replace(/(^-|-$)/g, '');
  }
  this.updatedAt = new Date();
  next();
});

// Instance method to add moderator
spaceSchema.methods.addModerator = function(userId, addedByUserId) {
  // Check if user is already a moderator
  const isModerator = this.moderators.some(mod => mod.user.toString() === userId.toString());
  if (isModerator) return false;

  this.moderators.push({
    user: userId,
    addedBy: addedByUserId
  });

  return this.save();
};

// Instance method to remove moderator
spaceSchema.methods.removeModerator = function(userId) {
  this.moderators = this.moderators.filter(mod => mod.user.toString() !== userId.toString());
  return this.save();
};

// Instance method to check if user is moderator
spaceSchema.methods.isModerator = function(userId) {
  return this.moderators.some(mod => mod.user.toString() === userId.toString());
};

// Instance method to check if user is owner
spaceSchema.methods.isOwner = function(userId) {
  return this.owner.toString() === userId.toString();
};

// Instance method to check if user can moderate
spaceSchema.methods.canModerate = function(userId) {
  return this.isOwner(userId) || this.isModerator(userId);
};

// Instance method to update stats
spaceSchema.methods.updateStats = function(field, increment = 1) {
  if (this.stats[field] !== undefined) {
    this.stats[field] += increment;
  }
  return this.save();
};

// Instance method to increment member count
spaceSchema.methods.incrementMemberCount = function() {
  this.memberCount += 1;
  return this.save();
};

// Instance method to decrement member count
spaceSchema.methods.decrementMemberCount = function() {
  this.memberCount = Math.max(0, this.memberCount - 1);
  return this.save();
};

// Instance method to increment post count
spaceSchema.methods.incrementPostCount = function(type = 'post') {
  this.postCount += 1;
  if (type === 'question') {
    this.questionCount += 1;
  } else if (type === 'discussion') {
    this.discussionCount += 1;
  }
  return this.save();
};

// Instance method to decrement post count
spaceSchema.methods.decrementPostCount = function(type = 'post') {
  this.postCount = Math.max(0, this.postCount - 1);
  if (type === 'question') {
    this.questionCount = Math.max(0, this.questionCount - 1);
  } else if (type === 'discussion') {
    this.discussionCount = Math.max(0, this.discussionCount - 1);
  }
  return this.save();
};

// Static method to find public spaces
spaceSchema.statics.findPublic = function() {
  return this.find({ isPublic: true, isActive: true });
};

// Static method to search spaces
spaceSchema.statics.search = function(query, limit = 10) {
  return this.find({
    $and: [
      { isPublic: true, isActive: true },
      {
        $or: [
          { name: { $regex: query, $options: 'i' } },
          { description: { $regex: query, $options: 'i' } }
        ]
      }
    ]
  })
  .sort({ memberCount: -1, postCount: -1 })
  .limit(limit);
};

// Static method to get trending spaces
spaceSchema.statics.getTrending = function(limit = 10) {
  const oneWeekAgo = new Date();
  oneWeekAgo.setDate(oneWeekAgo.getDate() - 7);

  return this.aggregate([
    {
      $match: {
        isPublic: true,
        isActive: true,
        lastActivity: { $gte: oneWeekAgo }
      }
    },
    {
      $addFields: {
        activityScore: {
          $add: [
            { $multiply: ['$memberCount', 0.1] },
            { $multiply: ['$postCount', 0.5] },
            { $multiply: ['$stats.totalViews', 0.01] }
          ]
        }
      }
    },
    {
      $sort: { activityScore: -1 }
    },
    {
      $limit: limit
    }
  ]);
};

module.exports = mongoose.model('Space', spaceSchema);
